# <center> Exploratory data analysis with Pandas

## 1. Demonstration of main Pandas methods

**[Pandas](http://pandas.pydata.org)** is a Python library that provides extensive means for data analysis. Data scientists often work with data stored in table formats like `.csv`, `.tsv`, or `.xlsx`. Pandas makes it very convenient to load, process, and analyze such tabular data using SQL-like queries. In conjunction with other useful library `Matplotlib` and `Seaborn`, `Pandas` provides a wide range of opportunities for visual analysis of tabular data.

The main data structures in `Pandas` are implemented with **DataFrame** classes. `DataFrames` are great for representing real data: rows correspond to instances (examples, observations, etc.), and columns correspond to features of these instances.

To run each cell we press **Shift+Enter**. Running the first cell should print "Executing the first cell of my class."

In [1]:
print ("Executing the first cell of my class")

Executing the first cell of my class


Next we import python libraries `numpy` and `pandas`. To do so, we execute the below cell

In [2]:
import numpy as np
import pandas as pd
pd.set_option("display.precision", 2)
pd.set_option('display.max_columns', None)

In this assignment, we'll demonstrate the main methods in action by analyzing a [dataset](https://bigml.com/user/francisco/gallery/dataset/5163ad540c0b5e5b22000383) on the churn rate of telecom operator clients. Let's read the data (using `read_csv`), and take a look at the first 5 lines using the `head` method:

In [3]:
df = pd.read_csv('telecom_churn.csv')
df.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  total eve minutes  total eve calls  total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

Recall that each row corresponds to one client, an **instance**, and columns are **features** of this instance.

Let’s have a look at data dimensionality, feature names, and feature types.

In [4]:
print(df.shape)

(3333, 21)


From the output, we can see that the table contains 3333 rows and 21 columns.

Now let's try printing out column names using `columns`:

In [5]:
print(list(df.columns))

['state', 'account length', 'area code', 'phone number', 'international plan', 'voice mail plan', 'number vmail messages', 'total day minutes', 'total day calls', 'total day charge', 'total eve minutes', 'total eve calls', 'total eve charge', 'total night minutes', 'total night calls', 'total night charge', 'total intl minutes', 'total intl calls', 'total intl charge', 'customer service calls', 'churn']


We can use the `info()` method to output some general information about the dataframe. Note that the 'object' data type refers to categorical variables in this data set. '3333 non-null' means that all 3333 values of the corresponding variable are not missing.  

In [6]:
print(df.info())   

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3333 non-null   object 
 1   account length          3333 non-null   int64  
 2   area code               3333 non-null   int64  
 3   phone number            3333 non-null   object 
 4   international plan      3333 non-null   object 
 5   voice mail plan         3333 non-null   object 
 6   number vmail messages   3333 non-null   int64  
 7   total day minutes       3333 non-null   float64
 8   total day calls         3333 non-null   int64  
 9   total day charge        3333 non-null   float64
 10  total eve minutes       3333 non-null   float64
 11  total eve calls         3333 non-null   int64  
 12  total eve charge        3333 non-null   float64
 13  total night minutes     3333 non-null   float64
 14  total night calls       3333 non-null   

`bool`, `int64`, `float64` and `object` are the data types of our features. We see that one feature is logical (`bool`), 3 features are of type `object`, and 16 features are numeric. With this same method, we can easily see if there are any missing values. Here, there are none because each column contains 3333 observations, the same number of rows we saw before with `shape`.

The `describe` method shows basic statistical characteristics of each numerical feature (`int64` and `float64` types): number of non-missing values, mean, standard deviation, range, median, 0.25 and 0.75 quartiles.

In [7]:
df.describe()

account length  area code  number vmail messages  total day minutes  \
count         3333.00    3333.00                3333.00            3333.00   
mean           101.06     437.18                   8.10             179.78   
std             39.82      42.37                  13.69              54.47   
min              1.00     408.00                   0.00               0.00   
25%             74.00     408.00                   0.00             143.70   
50%            101.00     415.00                   0.00             179.40   
75%            127.00     510.00                  20.00             216.40   
max            243.00     510.00                  51.00             350.80   

       total day calls  total day charge  total eve minutes  total eve calls  \
count          3333.00           3333.00            3333.00          3333.00   
mean            100.44             30.56             200.98           100.11   
std              20.07              9.26              50.71            19.92   
min               0.00              0.00               0.00             0.00   
25%              87.00             24.43             166.60            87.00   
50%             101.00             30.50             201.40           100.00   
75%             114.00             36.79             235.30           114.00   
max             165.00             59.64             363.70           170.00   

       total eve charge  total night minutes  total night calls  \
count           3333.00              3333.00            3333.00   
mean              17.08               200.87             100.11   
std                4.31                50.57              19.57   
min                0.00                23.20              33.00   
25%               14.16               167.00              87.00   
50%               17.12               201.20             100.00   
75%               20.00               235.30             113.00   
max               30.91               395.00             175.00   

       total night charge  total intl minutes  total intl calls  \
count             3333.00             3333.00           3333.00   
mean                 9.04               10.24              4.48   
std                  2.28                2.79              2.46   
min                  1.04                0.00              0.00   
25%                  7.52                8.50              3.00   
50%                  9.05               10.30              4.00   
75%                 10.59               12.10              6.00   
max                 17.77               20.00             20.00   

       total intl charge  customer service calls  
count            3333.00                 3333.00  
mean                2.76                    1.56  
std                 0.75                    1.32  
min                 0.00                    0.00  
25%                 2.30                    1.00  
50%                 2.78                    1.00  
75%                 3.27                    2.00  
max                 5.40                    9.00

In order to see statistics on non-numerical features, one has to explicitly indicate data types of interest in the `include` parameter.

In [8]:
df.describe(include=['object', 'bool'])

state phone number international plan voice mail plan  churn
count   3333         3333               3333            3333   3333
unique    51         3333                  2               2      2
top       WV     401-3926                 no              no  False
freq     106            1               3010            2411   2850

For categorical (type `object`) and boolean (type `bool`) features we can use the `value_counts` method. Let's have a look at the distribution of `churn`:

In [9]:
df['churn'].value_counts()

False    2850
True      483
Name: churn, dtype: int64

2850 users out of 3333 are *loyal*; their `Churn` value is `0`. To calculate fractions, pass `normalize=True` to the `value_counts` function.

In [10]:
df['churn'].value_counts(normalize=True)

False    0.86
True     0.14
Name: churn, dtype: float64


### Sorting

A DataFrame can be sorted by the value of one of the variables (i.e columns). For example, we can sort by *Total day charge* (use `ascending=False` to sort in descending order):


In [13]:
df.sort_values(by='total day charge', ascending=False).head(10)

state  account length  area code phone number international plan  \
365     CO             154        415     343-5709                 no   
985     NY              64        415     345-9140                yes   
2594    OH             115        510     348-1163                yes   
156     OH              83        415     370-9116                 no   
605     MO             112        415     373-2053                 no   
975     DE             129        510     332-6181                 no   
15      NY             161        415     351-7269                 no   
1679    OK              82        510     393-4823                 no   
2028    SD              93        510     408-4836                 no   
197     TX             208        510     378-3625                 no   

     voice mail plan  number vmail messages  total day minutes  \
365               no                      0              350.8   
985               no                      0              346.8   
2594              no                      0              345.3   
156               no                      0              337.4   
605               no                      0              335.5   
975               no                      0              334.3   
15                no                      0              332.9   
1679              no                      0              329.8   
2028              no                      0              328.1   
197               no                      0              326.5   

      total day calls  total day charge  total eve minutes  total eve calls  \
365                75             59.64              216.5               94   
985                55             58.96              249.5               79   
2594               81             58.70              203.4              106   
156               120             57.36              227.4              116   
605                77             57.04              212.5              109   
975               118             56.83              192.1              104   
15                 67             56.59              317.8               97   
1679               73             56.07              208.3              120   
2028              106             55.78              151.7               89   
197                67             55.51              176.3              113   

      total eve charge  total night minutes  total night calls  \
365              18.40                253.9                100   
985              21.21                275.4                102   
2594             17.29                217.5                107   
156              19.33                153.9                114   
605              18.06                265.0                132   
975              16.33                191.0                 83   
15               27.01                160.6                128   
1679             17.71                267.1                102   
2028             12.89                303.5                114   
197              14.99                181.7                102   

      total night charge  total intl minutes  total intl calls  \
365                11.43                10.1                 9   
985                12.39                13.3                 9   
2594                9.79                11.8                 8   
156                 6.93                15.8                 7   
605                11.93                12.7                 8   
975                 8.59                10.4                 6   
15                  7.23                 5.4                 9   
1679               12.02                10.6                 6   
2028               13.66                 8.7                 3   
197                 8.18                10.7                 6   

      total intl charge  customer service calls  churn  
365                2.73                       1   True  
985                3.59             

Sorting Total day charge in ascending order

In [14]:
df.sort_values(by='total day charge').head(10)

state  account length  area code phone number international plan  \
1345    SD              98        415     392-2555                 no   
1397    VT             101        510     413-7655                 no   
2736    OK             127        510     403-1128                 no   
2753    OH             134        415     406-4158                 no   
1986    WI              70        415     405-9233                 no   
1052    OR              98        415     378-6772                yes   
2252    NH             148        408     333-7449                 no   
3046    MI             110        510     357-5784                 no   
1621    SC             138        510     370-9533                 no   
1076    WY              53        415     337-4339                 no   

     voice mail plan  number vmail messages  total day minutes  \
1345              no                      0                0.0   
1397              no                      0                0.0   
2736             yes                     27                2.6   
2753              no                      0                7.8   
1986              no                      0                7.9   
1052              no                      0               12.5   
2252              no                      0               17.6   
3046              no                      0               18.9   
1621             yes                     21               19.5   
1076             yes                     27               25.9   

      total day calls  total day charge  total eve minutes  total eve calls  \
1345                0              0.00              159.6              130   
1397                0              0.00              192.1              119   
2736              113              0.44              254.0              102   
2753               86              1.33              171.4              100   
1986              100              1.34              136.4               83   
1052               67              2.13              256.6               90   
2252              121              2.99              161.7              125   
3046               92              3.21              258.4               81   
1621              149              3.32              140.9              109   
1076              119              4.40              206.5               96   

      total eve charge  total night minutes  total night calls  \
1345             13.57                167.1                 88   
1397             16.33                168.8                 95   
2736             21.59                242.7                156   
2753             14.57                186.5                 80   
1986             11.59                156.6                 89   
1052             21.81                169.4                 88   
2252             13.74                203.1                 82   
3046             21.96                109.6                 74   
1621             11.98                179.7                111   
1076             17.55                228.1                 64   

      total night charge  total intl minutes  total intl calls  \
1345                7.52                 6.8                 1   
1397                7.60                 7.2                 4   
2736               10.92                 9.2                 5   
2753                8.39                12.9                 2   
1986                7.05                12.1                 1   
1052                7.62                 7.7                 9   
2252                9.14                10.6                 6   
3046                4.93                14.8                 4   
1621                8.09                 7.9                 1   
1076               10.26                 6.5                 7   

      total intl charge  customer service calls  churn  
1345               1.84                       4   True  
1397               1.94             

### Indexing and retrieving data

A DataFrame can be indexed in a few different ways. 

To get a single column/feature, you can use a `DataFrame['Name']` construction. Let's use this to answer a question about that column alone: **what is the proportion of churned users in our dataframe?**

In [15]:
df['churn'].mean()

0.14491449144914492

14.5% is actually quite bad for a company; such a churn rate can make the company go bankrupt.

**Boolean indexing** with one column is also very convenient. The syntax is `df[df['Name'] == P]`, where `P` is some logical condition that is checked for each element of the `Name` column. The result of such indexing is the DataFrame consisting only of rows that satisfy the `P` condition on the `Name` column. 

Lets use it to see all customers who have churned out of the company **('churn' == 1)**


In [18]:
df[df['churn'] == True]

state  account length  area code phone number international plan  \
10      IN              65        415     329-6603                 no   
15      NY             161        415     351-7269                 no   
21      CO              77        408     393-7984                 no   
33      AZ              12        408     360-1596                 no   
41      MD             135        408     383-6029                yes   
...    ...             ...        ...          ...                ...   
3301    CA              84        415     417-1488                 no   
3304    IL              71        510     330-7137                yes   
3320    GA             122        510     411-5677                yes   
3322    MD              62        408     409-1856                 no   
3323    IN             117        415     362-5899                 no   

     voice mail plan  number vmail messages  total day minutes  \
10                no                      0              129.1   
15                no                      0              332.9   
21                no                      0               62.4   
33                no                      0              249.6   
41               yes                     41              173.1   
...              ...                    ...                ...   
3301              no                      0              280.0   
3304              no                      0              186.1   
3320              no                      0              140.0   
3322              no                      0              321.1   
3323              no                      0              118.4   

      total day calls  total day charge  total eve minutes  total eve calls  \
10                137             21.95              228.5               83   
15                 67             56.59              317.8               97   
21                 89             10.61              169.9              121   
33                118             42.43              252.4              119   
41                 85             29.43              203.9              107   
...               ...               ...                ...              ...   
3301              113             47.60              202.2               90   
3304              114             31.64              198.6              140   
3320              101             23.80              196.4               77   
3322              105             54.59              265.5              122   
3323              126             20.13              249.3               97   

      total eve charge  total night minutes  total night calls  \
10               19.42                208.8                111   
15               27.01                160.6                128   
21               14.44                209.6                 64   
33               21.45                280.2                 90   
41               17.33                122.2                 78   
...                ...                  ...                ...   
3301             17.19                156.8                103   
3304             16.88                206.5                 80   
3320             16.69                120.1                133   
3322             22.57                180.5                 72   
3323             21.19                227.0                 56   

      total night charge  total intl minutes  total intl calls  \
10                  9.40                12.7                 6   
15                  7.23                 5.4                 9   
21                  9.43                 5.7                 6   
33                 12.61                11.8                 3   
41                  5.50                14.6                15   
...                  ...                 ...               ...   
3301                7.06                10.4                 4   
3304                9.29                13.8                 5   

Let's answer the question:

**What are average values of numerical features for churned users?**

In [19]:
df[df['churn'] == True].mean()

account length            102.66
area code                 437.82
number vmail messages       5.12
total day minutes         206.91
total day calls           101.34
total day charge           35.18
total eve minutes         212.41
total eve calls           100.56
total eve charge           18.05
total night minutes       205.23
total night calls         100.40
total night charge          9.24
total intl minutes         10.70
total intl calls            4.16
total intl charge           2.89
customer service calls      2.23
churn                       1.00
dtype: float64

**What are average values of numerical features for non-churned users?**

In [20]:
df[df['churn'] == False].mean()

account length            100.79
area code                 437.07
number vmail messages       8.60
total day minutes         175.18
total day calls           100.28
total day charge           29.78
total eve minutes         199.04
total eve calls           100.04
total eve charge           16.92
total night minutes       200.13
total night calls         100.06
total night charge          9.01
total intl minutes         10.16
total intl calls            4.53
total intl charge           2.74
customer service calls      1.45
churn                       0.00
dtype: float64

**How much time (on average) do churned users spend on the phone during daytime?**

In [21]:
df[df['churn'] == True]['total day minutes'].mean()

206.91407867494823


**What is the maximum length of international calls among loyal users (`churn == 0`) who do not have an international plan?**



In [24]:
df[(df['churn'] == False) & (df['international plan'] == 'no')]['total intl minutes'].max()

18.9


### Grouping

In general, grouping data in Pandas works as follows:



```python
df.groupby(by=grouping_columns)[columns_to_show].function()
```


1. First, the `groupby` method divides the `grouping_columns` by their values. They become a new index in the resulting dataframe.
2. Then, columns of interest are selected (`columns_to_show`). If `columns_to_show` is not included, all non groupby clauses will be included.
3. Finally, one or several functions are applied to the obtained groups per selected columns.

Here is an example where we group the data according to the values of the `Churn` variable and display statistics of three columns in each group:

In [25]:
columns_to_show = ['total day minutes', 'total eve minutes', 'total night minutes']

df.groupby(['churn'])[columns_to_show].describe(percentiles=[])

total day minutes                                   total eve minutes  \
                  count    mean    std  min    50%    max             count   
churn                                                                         
False            2850.0  175.18  50.18  0.0  177.2  315.6            2850.0   
True              483.0  206.91  69.00  0.0  217.6  350.8             483.0   

                                         total night minutes                 \
         mean    std   min    50%    max               count    mean    std   
churn                                                                         
False  199.04  50.29   0.0  199.6  361.8              2850.0  200.13  51.11   
True   212.41  51.73  70.9  211.3  363.7               483.0  205.23  47.13   

                            
        min     50%    max  
churn                       
False  23.2  200.25  395.0  
True   47.4  204.80  354.9


### Summary tables

Suppose we want to see how the observations in our sample are distributed in the context of two variables - `churn` and `international plan`. To do so, we can build a **contingency table** using the `crosstab` method:



In [51]:
pd.crosstab(df['churn'], df['international plan'])
def cond_churn_probability(churn, intl):
    p_of_intl_plan = len(df[df['international plan']== intl])/(len(df))
    p_of_churn_given_intl= 

0.0969096909690969


In [46]:
pd.crosstab(df['churn'], df['voice mail plan'])

voice mail plan    no  yes
churn                     
False            2008  842
True              403   80


### DataFrame transformations

Like many other things in Pandas, adding columns to a DataFrame can be performed in many ways.

For example, if we want to calculate the total charge of calls for all users, let's create the `total charge` Series and paste it into the DataFrame:



Notice the last column "total charge," which we have added in our data frame 

In [ ]:
df['total charge'] = df['total day charge'] + df['total eve charge'] + \
                     df['total night charge'] + df['total intl charge']
df.head()

To delete columns or rows, use the `drop` method, passing the required indexes and the `axis` parameter (`1` if you delete columns, and nothing or `0` if you delete rows). The `inplace` argument tells whether to change the original DataFrame. With `inplace=False`, the `drop` method doesn't change the existing DataFrame and returns a new one with dropped rows or columns. With `inplace=True`, it alters the DataFrame.

In [ ]:
# get rid of just created column and phone number                
df.drop(['total charge', 'phone number'], axis=1, inplace=True) 
# and here’s how you can delete rows
#df.drop([1, 2]).head() 

## 2. First attempt at predicting telecom churn


Let's see how churn rate is related to the *International plan* feature. We'll do this using a `crosstab` contingency table and also through visual analysis with `Seaborn`


In [ ]:
pd.crosstab(df['churn'], df['international plan'], margins=True)

In [ ]:
# some imports to set up plotting 
import matplotlib.pyplot as plt
# pip install seaborn 
import seaborn as sns
# Graphics in retina format are sharper and more legible
%config InlineBackend.figure_format = 'retina'

In [ ]:
sns.countplot(x='international plan', hue='churn', data=df);


We see that, with *International Plan*, the churn rate is much higher, which is an interesting observation! Perhaps large and poorly controlled expenses with international calls are very conflict-prone and lead to dissatisfaction among the telecom operator's customers.

Next, let's look at another important feature – *Customer service calls*. Let's also make a summary table and a picture.

In [ ]:
pd.crosstab(df['churn'], df['customer service calls'], margins=True)

In [ ]:
sns.countplot(x='customer service calls', hue='churn', data=df);

Although it's not so obvious from the summary table, it's easy to see from the above plot that the churn rate increases sharply from 4 customer service calls and above.

Now let's add a binary feature to our DataFrame – `Customer service calls > 3`. And once again, let's see how it relates to churn. 

In [ ]:
df['many_service_calls'] = (df['customer service calls'] > 3).astype('int')

pd.crosstab(df['many_service_calls'], df['churn'], margins=True)

In [ ]:
sns.countplot(x='many_service_calls', hue='churn', data=df);


Let's construct another contingency table that relates *Churn* with both *International plan* and freshly created *Many_service_calls*.



In [ ]:
pd.crosstab(df['many_service_calls'] & (df['international plan'] == 'yes') , df['churn'], margins=True)

Therefore, predicting that a customer is not loyal (*Churn*=1) in the case when the number of calls to the service center is greater than 3 and the *International Plan* is added (and predicting *Churn*=0 otherwise), we might expect an accuracy of 85.8% (we are mistaken only 464 + 9 times). This number, 85.8%, that we got through this very simple reasoning serves as a good starting point (*baseline*) for the further machine learning models that we will build. 

Recall that, before the advent of machine learning, the data analysis process looked something like this. Let's recap what we've covered:
    
- The share of loyal clients in the sample is 85.5%. The most naive model that always predicts a "loyal customer" on such data will guess right in about 85.5% of all cases. That is, the proportion of correct answers (*accuracy*) of subsequent models should be no less than this number, and will hopefully be significantly higher;
- With the help of a simple forecast that can be expressed by the following formula: "International plan = True & Customer Service calls > 3 => Churn = 1, else Churn = 0", we can expect a guessing rate of 85.8%, which is just above 85.5%. 
- We got these two baselines without applying machine learning, and they'll serve as the starting point for our subsequent models. 

This notebook is adaption from the following kaggle notebook https://www.kaggle.com/kashnitsky/topic-1-exploratory-data-analysis-with-pandas. 